src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -149,6 +148,297 @@ public interface IFileWatchingToken
+
+            public static WatcherOperation WatchDirectory(string directory, string? filter, IVsFreeThreadedFileChangeEvents2 sink, List<uint> cookies)
+                => new(Kind.WatchDirectory, directory, filter, sink, cookies);
+
+            public static WatcherOperation WatchFile(string path, _VSFILECHANGEFLAGS fileChangeFlags, IVsFreeThreadedFileChangeEvents2 sink, Context.FileWatchingToken token)
+                => new(Kind.WatchFile, path, fileChangeFlags, sink, token);
+
+            public static WatcherOperation UnwatchDirectories(List<uint> cookies)
+                => new(Kind.UnwatchDirectories, cookies);
+
+            public static WatcherOperation UnwatchFiles(IEnumerable<Context.FileWatchingToken> tokens)
+                => new(Kind.UnwatchFiles, tokens);
+
+            public static WatcherOperation UnwatchFile(Context.FileWatchingToken token)
+                => new(Kind.UnwatchFile, token);
+
+            /// <summary>
+            /// Attempts to combine the current <see cref="WatcherOperation"/> with the next operation in sequence. When
+            /// successful, <paramref name="combined"/> is assigned a value which, when applied, performs an operation
+            /// equivalent to performing the current instance immediately followed by <paramref name="other"/>.
+            /// </summary>
+            /// <param name="other">The next operation to apply.</param>
+            /// <param name="combined">An operation representing the combined application of the current instance and
+            /// <paramref name="other"/>, in that order; otherwise, <see cref="Empty"/> if the current operation cannot
+            /// be combined with <paramref name="other"/>.</param>
+            /// <returns><see langword="true"/> if the current operation can be combined with <paramref name="other"/>;
+            /// otherwise, <see langword="false"/>.</returns>
+            public bool TryCombineWith(in WatcherOperation other, out WatcherOperation combined)
+            {
+                if (other._kind == Kind.None)
+                {
+                    combined = this;
+                    return true;
+                }
+                else if (_kind == Kind.None)
+                {
+                    combined = other;
+                    return true;
+                }
+
+                switch (_kind)
+                {
+                    case Kind.WatchDirectory:
+                    case Kind.WatchFile:
+                        // Watching operations cannot be combined
+                        break;
+
+                    case Kind.UnwatchFile when other._kind == Kind.UnwatchFile:
+                        combined = UnwatchFiles(ImmutableList.Create(_token, other._token));
+                        return true;
+
+                    case Kind.UnwatchFile when other._kind == Kind.UnwatchFiles:
+                        combined = UnwatchFiles(other._tokens.ToImmutableList().Insert(0, _token));
```

Comment:
i don't love that a run of N events would create all those immutablelist forks.  part of me thinks it would be nice to just walk all the events and merge into a temp buffer that doesn't cause allocation forking.  but it's not a big deal.

Reply:
It's not quite as inefficient as it looks. If `_tokens` is already an immutable list, `ToImmutableList()` will return that instance unchanged. The first merge step will construct this list, but the second and subsequent will just mutate it.

Reply:
right.  i was referring more to the second N steps continually making new immurtable lists.  but i'm fine doing this and only adjuting if we run into any perf issues in the future.